backend/src/Edna/Upload/Service.hs

```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+  FileContents -> Edna FileSummary
+uploadFile' projSqlId@(SqlId proj) methodSqlId@(SqlId method)
+  description fileName fileBytes fc = do
+    let projId = ProjectId $ fromIntegral proj
+    let methodId = TestMethodologyId $ fromIntegral method
+    runSelectReturningOne' (lookup_ (esProject ednaSchema) projId)
+      `whenNothingM_`
+      throwM (UEUnknownProject projSqlId)
+    runSelectReturningOne' (lookup_ (esTestMethodology ednaSchema) methodId)
+      `whenNothingM_`
+      throwM (UEUnknownTestMethodology methodSqlId)
+
+    transact $ insertAll (fromIntegral proj) (fromIntegral method)
+  where
+    insertAll :: HasCallStack => Word32 -> Word32 -> Edna FileSummary
+    insertAll projId methodId = do
+      let fileMeasurements = fcMeasurements fc
+      targetToId <- HM.fromList <$> mapM insertTarget (keys fileMeasurements)
+      let compounds = hashNub $ flip concatMap (elems fileMeasurements) $
+            \(TargetMeasurements targetMeasurements) -> keys targetMeasurements
+      compoundToId <- HM.fromList <$> mapM insertCompound compounds
+
+      expFileId <- insertExperimentFile projId methodId (fcMetadata fc)
+        description fileName fileBytes
+
+      forM_ (toPairs fileMeasurements) $
+        \(targetName, TargetMeasurements targetMeasurements) ->
+          forM_ (toPairs targetMeasurements) $ \(compoundName, measurements) -> do
+            let
+              getId :: HasCallStack => Text -> HashMap Text Word32 -> Word32
+              getId name =
+                fromMaybe (error $ "unexpected name: " <> name) .
+                view (at name)
+            let targetId = getId targetName targetToId
+            let compoundId = getId compoundName compoundToId
+            insertExperiment expFileId compoundId targetId measurements
+
+      measurementsToSummary fileMeasurements
+
+insertTarget :: Text -> Edna (Text, Word32)
+insertTarget targetName = (targetName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esTarget ednaSchema)
+    (insertExpressions [TargetRec default_ (val_ targetName) default_])
+    (Pg.onConflict (Pg.conflictingFields tName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added target not found: " <> targetName)  <$>
+    targetNameToId targetName
+
+insertCompound :: Text -> Edna (Text, Word32)
+insertCompound compoundName = (compoundName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esCompound ednaSchema)
+    (insertExpressions [CompoundRec default_ (val_ compoundName) default_ default_])
+    (Pg.onConflict (Pg.conflictingFields cName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added compound not found: " <> compoundName)  <$>
+    compoundNameToId compoundName
+
+insertExperimentFile ::
+  Word32 -> Word32 -> FileMetadata -> Text -> Text -> LByteString -> Edna Word32
+insertExperimentFile projId methodId meta descr fileName blob = do
+  unSerial . expectOneInsertion . map efExperimentFileId <$> runInsertReturningList'
```

Comment:
Interesting, I don't know why I did it this way, probably because there is no `runInsertReturningOne'` in `Edna.DB.Integration`, but I guess I can add it.

Reply:
Apparently there is no `runInsertReturningOne` in `beam`. But I still can define it.

Reply:
There is `runInsertReturning`, isn't it what you want?

Reply:
Looks like it isn't (

Reply:
It looks a little bit different. Anyway, I've defined
```
-- | Insert items and expect one item in the result. The caller is responsible
-- for ensuring that exactly one result will be returned.
runInsertReturningOne' ::
  ( Beamable table
  , FromBackendRow Postgres (table Identity)
  ) => SqlInsert Postgres table -> Edna (table Identity)
runInsertReturningOne' = fmap expectOneInsertion . runInsertReturningList'
  where
    expectOneInsertion :: HasCallStack => [x] -> x
    expectOneInsertion = \case
      [x] -> x
      xs -> error $
        "Expected to insert 1 item, but inserted: " <> show (length xs)
```

and it's probably sufficient for now